src/package_json_serde.rs

```diff
@@ -0,0 +1,337 @@
+                json_object.get("type").and_then(|ty| serde_json::from_value(ty.clone()).ok());
+            package_json.side_effects = json_object.get("sideEffects").cloned();
+        }
+
+        package_json.path = path;
+        package_json.realpath = realpath;
+        package_json.raw_json = std::sync::Arc::new(raw_json);
+        Ok(package_json)
+    }
+
+    fn get_value_by_path<'a>(
+        fields: &'a serde_json::Map<String, JSONValue>,
+        path: &[String],
+    ) -> Option<&'a JSONValue> {
+        if path.is_empty() {
+            return None;
+        }
+        let mut value = fields.get(&path[0])?;
+        for key in path.iter().skip(1) {
+            if let Some(inner_value) = value.as_object().and_then(|o| o.get(key)) {
+                value = inner_value;
+            } else {
+                return None;
+            }
+        }
+        Some(value)
+    }
+
+    /// Raw serde json value of `package.json`.
+    ///
+    /// This is currently used in Rspack for:
+    /// * getting the `sideEffects` field
+    /// * query in <https://www.rspack.dev/config/module.html#ruledescriptiondata> - search on GitHub indicates query on the `type` field.
+    ///
+    /// To reduce overall memory consumption, large fields that useless for pragmatic use are removed.
+    /// They are: `description`, `keywords`, `scripts`,
+    /// `dependencies` and `devDependencies`, `peerDependencies`, `optionalDependencies`.
+    #[cfg(feature = "package_json_raw_json_api")]
+    #[must_use]
+    pub const fn raw_json(&self) -> &std::sync::Arc<JSONValue> {
+        &self.raw_json
+    }
+
+    /// The "browser" field is provided by a module author as a hint to javascript bundlers or component tools when packaging modules for client side use.
+    /// Multiple values are configured by [ResolveOptions::alias_fields].
+    ///
+    /// <https://github.com/defunctzombie/package-browser-field-spec>
+    fn browser_fields<'a>(
+        &'a self,
+        alias_fields: &'a [Vec<String>],
+    ) -> impl Iterator<Item = &'a JSONMap> + 'a {
+        alias_fields.iter().filter_map(|object_path| {
+            self.raw_json
+                .as_object()
+                .and_then(|json_object| Self::get_value_by_path(json_object, object_path))
+                // Only object is valid, all other types are invalid
+                // https://github.com/webpack/enhanced-resolve/blob/3a28f47788de794d9da4d1702a3a583d8422cd48/lib/AliasFieldPlugin.js#L44-L52
+                .and_then(|value| value.as_object())
+        })
+    }
+
+    fn alias_value<'a>(key: &Path, value: &'a JSONValue) -> Result<Option<&'a str>, ResolveError> {
+        match value {
+            JSONValue::String(value) => Ok(Some(value.as_str())),
+            JSONValue::Bool(b) if !b => Err(ResolveError::Ignored(key.to_path_buf())),
+            _ => Ok(None),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct ImportsExportsSerdeEntry<'a>(pub(crate) &'a JSONValue);
```

Comment:
I was a bit bumped myself about the boilerplate needed for implementing these `ImportsExports*` traits. Initially I had tried a different approach with concrete types, where I mapped the `JSONValue`s to the concrete types eagerly. Unfortunately, it didn't only require additional allocations, but it also subtly broke a few tests due to the nature of the eager evaluation. So I took the hard road instead :)